### ISSUE TITLE
Add 'detach' mode for `watch`

### ISSUE THREAD

**Issue by JefvdA** [docker/compose#11344]
### Description

Hi! :wave: 

I've mainly been using `docker compose up -d` when upping my compose services in the background. 
Recently, I've been wanting to move from `up` to `watch` as I would love to experiment with `watch` vs using `volumes`.

Intentionally I tried `docker compose watch -d` , only to find out it didn't have a detach mode like `up` does.

_____

Is this a feature that hasn't been added by choice, because the `watch` process needs to stay active to be able to sync/rebuild the containers?

Thanks!

**Comment by ndeloof** [comment]
You can combine `docker compose up -d` with `docker compose watch --no-up` to get the requested workflow

**Comment by JefvdA** [comment]
Hi @ndeloof , thanks for the quick response. 

Still, the watch process can't be kept in the background, meaning you'll still need to keep the terminal session open where you ran `docker compose watch --no-up` , to have files actually be synced.
_____

What I meant with my original comment was a `--detach` or `-d` flag for `watch` , so that the syncing of the files could be done in the background. Without having to keep the terminal sessions open.

Did that make sense ? :smile:

**Comment by ndeloof** [comment]
A command can't "run in the background", this is a shell feature to run command in the background and "detach". Just run `docker compose watch &`

**Comment by KevinMind** [comment]
I've tried this and end up with zombie processes. Has anyone implemented the suggested workflow with a way to prevent ongoing watch commands starting up docker compose after you want to bring down the project? doesn't seem like docker compose down kills the background watch process.

**Comment by adamlacoste** [comment]
+1 for this.  The recommendation to background the watch process is not a good solution.  Not only is the backgrounded process not terminated by docker compose down`, as KevinMind mentions, but attempting to kill the process manually can cause Docker to hang, forcing you to restart it.  And that is to say nothing of the hassle of having to manage STDOUT and STDERR from the backgrounded process.

What we need is an option or environment variable which can tell `docker compose watch` to keep synching/restarting services when the watched paths on the host system change, but otherwise run "detached" like `docker compose up -d` does.  Alternatively, an option for `docker compose up` like `--watch-detach` or `-W` would work fine as well.

**Comment by ndeloof** [comment]
@adamlacoste  `docker compose up --watch` is already implemented

**Comment by adamlacoste** [comment]
@ndeloof I hope this will not sound rude but it seems like you didn't read my comment fully.  Running `docker compose up -w` will attach output from my services to my terminal; cancelling out of that will kill my services.  I want to be able to have Docker watch my develop paths and sync their contents but I don't want the output attached to my terminal.  I would like this watch mode to be persistent, even through system restarts, until I deliberately end it (e.g. with `docker compose down`).

Right now `docker compose up -dw` specifically tells you that detach mode cannot be combined with watch mode.  There seems to be no elegant way to detach from the `docker compose watch` command while in continues to work behind the scenes, either.

**Comment by ndeloof** [comment]
`watch` can't be "persistent": it requires a process to monitor filesystem changes and apply restart/recreate policy. This process is the `docker compose watch` command you run, if you don't want it to block your terminal, just run it as a background process.

**Comment by YanisThePie** [comment]
the problem with the fact that we cannot use watch in detached mode has repercussions in several ways, in particular, it makes -watch command impossible to use with Visual Studio IDE and its debugger, because the -d command cannot be removed there.

**Comment by ndeloof** [comment]
@YanisThePie compose being a client side, stateless command line, a process need to be ran and stay to implement `watch`. I can't tell about Visual Studio IDE integration constraints, but I can imagine that's not an isolated need

**Comment by adamlacoste** [comment]
Obviously I don't know what's possible or impossible under the hood.  But from a user's perspective, it seems strange that I can run `docker compose up -d` to start (detached) service containers which are persistent even through system restarts, yet I can't simply add `-w` to that command to enable the watch functionality persistently as well.

Right now, the closest I can get to that behavior is by using `screen docker compose watch` and then detaching from the screen.  It does not survive system restarts but otherwise allows the watching to happen in the background and is easy to kill if I want to later on.

**Comment by ndeloof** [comment]
compose is a client binary, it let docker engine create and run containers and manage their lifecycle, which can include restarting after failure or reboot. just like `docker run ...` its role ends after engine accepted the container start request, so the ability to run "detached". `watch` on the other hand is a client-side feature which require to monitor filesystem. Docker engine has no access to client environment (can run on a distinct machine)

**Comment by atruskie** [comment]
Sorry to resurrect this, but after a lot of searching I found this issue. The context for this issue is useful. Perhaps it could be mentioned in the docs for watch?

**Comment by devnix** [comment]
>A command can't "run in the background", this is a shell feature to run command in the background and "detach". Just run docker compose watch &

I strongly disagree. `docker compose up -d` still allows you to close the terminal without killing the process. You still would need to look for more trickery like disowning the process or running it in other tools like `screen` or `tmux`.

As you can run `docker compose up -d`, I would expect `docker compose up -d --watch` to be compatible too 🙁

**Comment by ndeloof** [comment]
> docker compose up -d still allows you to close the terminal without killing the process

`up -d` connects to docker deamon over API, create a new container, then is done. It's not about running a background process or such "trickery", but just about the time compose is in charge for your app and need to keep running. `watch` requires a long-running process to monitor filesystem, so can't be "detached"

**Comment by xenago** [comment]
Yeah this is a much-needed improvement, would be excellent to have this added!

**Comment by a463** [comment]
This would actually make `--watch` usable in teams. We set it up in a project, but had to revert to volumes, as it quickly got annoying to ask every team member to run their docker compose with `--watch`, check the terminal often, and keep the process running. We were getting complaints they were not seeing changes after a `git pull`.

What's worse, after a `git pull`, running `docker compose up --watch` does not sync the changes, as the command is not aware any have occurred, so we had to explain that a rebuild needs to be triggered manually.

It would be nice to be able to run `--watch` just once in detached mode, and have docker compose somehow sync the files in the background. I'm not very familiar with the internals, so not sure if feasible or a huge undertaking, but perhaps `--watch --detach` could run a separate docker container (sidecar?), which mounts the local volume (i.e. `watch.path`) and the volume from the container (i.e. `watch.target`), and watches & sync files between them? If so, this might even become a default and syncing could "just work" in the future, without even requiring `--watch`.

**Comment by floyd-may** [comment]
> Sorry to resurrect this, but after a lot of searching I found this issue. The context in this issue is useful. Perhaps it could be mentioned in the docs for watch?

Agree. Specifically, the comparison between watch and bind mounts. Bind mounts do work "in the background" so to speak so clarity on that in the docs would definitely have saved me some time and headache.

**Comment by o-nix** [comment]
So, they are saying that `watch` is easier/better/granular/etc. than simple mounts, while it can not even run in the background?!
I am bringing the `volumes` back and forgetting about this "improvement" I wanted to experiment with.

**Comment by ndeloof** [comment]
@o-nix 
> watch is easier/better/granular/etc. than simple mounts, while it can not even run in the background?

Those are very different things. bind mount is a container create option, it relies on Linux kernel to let container access mounted filesystem, so requires engine to run on same machine (or Docker Desktop host-to-VM mount)
 
watch relies on docker API. It allows use of an arbitrary (remote) engine and can be configure with fine grain rules - typically, excluding some sub-paths). But it require a client-side process to monitor filesystem and propagate changes to container.

**Comment by o-nix** [comment]
@ndeloof thanks, good to know the reason behind, and possibly have the ability to deploy something somewhere non-local.
However, my point is that the whole section is named `develop`, so I think it should be more oriented towards the standard
> Run `docker compose up -d` or click a green triangle in the opened compose.yaml in your IntelliJ. Then navigate to http://localhost:... and check your fully working app. When you need changes, rebuild your JAR/React/etc. app, and reload the browser; all updated files will already be there."

nice and easily understandable workflow.
Not like
> ...and also don't forget to keep a dedicated console with the `watch` command". Beware that if a console session dies, you need to set everything up again.

**Comment by floyd-may** [comment]
My hope was that this would be a more reasonable solution to the workarounds I have to do developing on Windows and getting file watcher tools / HMR to work because of [this issue](https://github.com/microsoft/WSL/issues/4739). No such luck 😞
